Turn burner & electric press into proper pseudomultiblocks

abel1502 opened the pull request:
They seem to have been implemented largely based on copy-pasted furnace code, and seemingly predate `BlockDummyable`. That meant the top two blocks of a press were intangible. Now they're in line with the rest of the machines, including the conveyor press.

I also added a bit of compatibility handling to `BlockDummyable`. It can now identify legacy monoblocks when deleting a suspected orphan, and promote them to lone cores. This makes this change savegame-compatible. Existing presses will remain single-block-sized and functionally identical to before. If broken and replaced, they will be 3-block-tall. The solution is general (the updated block must override `isLegacyMonoblock` to detect if it is a correct legacy block. There is a caveat, though: the fix is only triggered when the block is updated, which doesn't happen on world load, unfortunately. The best workaround for this that I came up with was triggering a block update from the tile entity tick if the metadata for the block indicates it's a dummy. The performance cost is negligible, thankfully.

abel1502 committed:
Turn presses into pseudomultiblocks

Affects burner and electrical press. They seem to have been implemented largely based on copy-pasted furnace code, and seemingly predate BlockDummyable. That meant the top two blocks of a press were intangible. Now they're in line with the rest of the machines, including the conveyor press.

abel1502 committed:
Fix createNewTileEntity and clean up a bit

abel1502 committed:
More fixes

Turns out, pseudomultiblocks are pretty complicated

abel1502 committed:
Remove broken & redundant transformMeta

abel1502 committed:
Add automatic migration API to BlockDummyable

abel1502 committed:
Make new press implementation savegame-compatible

abel1502 committed:
Trigger the press fix automatically

abel1502 committed:
Fix press auto-migration

70000hp commented:
I think that by now them not being proper multiblocks is deliberate?

abel1502 commented:
@70000hp considering their code used mojang-obfuscated variable names, I'm pretty sure it just wasn't touched for a long while

abel1502 committed:
Simplify press click handling

Turns out there is already a function for doing just that

abel1502 commented:
I think there might be an issue with this, marking as a draft while I'm investigating. I just saw a lot of multiblocks have disappeared from my world. In all likelyness, it's probably a consequence of one of the intermediate attempts, considering it doesn't seem to happen anymore, but better safe than sorry. I'll undraft when I'm sure it's safe

abel1502 committed:
Prevent pseudomultiblocks disappearing on chunk boundaries

If this was indeed a problem, it must've been one before my PR... I suspect maybe some of my intermediate implementations allowed this to have an effect. Regardless, this is definitely a useful fix

abel1502 commented:
Okay, I think I found the problem. I'll test my fix ~~tomorrow~~ right now.

Basically, a dummyable on a chunk border might wrongly consider itself an orphan if its neighbor is in an unloaded chunk. It's weird that the problem didn't show up before, as it was present before my changes. I suspect one of my intermediate solutions must've aggravated its effect.

abel1502 commented:
Okay, yes, I can confirm the problem no longer occurs. Removing the draft status

MellowArpeggiation commented:
vetoing this one, removes ability to squish PU bobblehead and plushies

70000hp commented:
> vetoing this one, removes ability to squish PU bobblehead and plushies

seconded

abel1502 commented:
> vetoing this one, removes ability to squish PU bobblehead and plushies

Compelling argument. Might make it unmultiblockable with a handdrill or something

abel1502 committed:
Make presses un-multiblockable with a hand drill

As per @MellowArpeggiation 's request. Meant for schenanigans with visually squishing things

abel1502 commented:
@MellowArpeggiation @70000hp here ya go

abel1502 committed:
Don't break the whole thing
